pkg/cloudprovider/aws/aws.go

```diff
@@ -506,3 +824,355 @@ func (self *AWSCloud) GetZone() (cloudprovider.Zone, error) {
+		return false, fmt.Errorf("multiple security groups found with same id")
+	}
+	group := groups[0]
+
+	newPermissions := []ec2.IPPerm{}
+
+	for _, port := range ports {
+		found := false
+		for _, permission := range group.IPPerms {
+			if permission.FromPort != port {
+				continue
+			}
+			if permission.ToPort != port {
+				continue
+			}
+			if permission.Protocol != protocol {
+				continue
+			}
+			if len(permission.SourceIPs) != 1 {
+				continue
+			}
+			if permission.SourceIPs[0] != sourceIp {
+				continue
+			}
+			found = true
+			break
+		}
+
+		if !found {
+			newPermission := ec2.IPPerm{}
+			newPermission.FromPort = port
+			newPermission.ToPort = port
+			newPermission.SourceIPs = []string{sourceIp}
+			newPermission.Protocol = protocol
+
+			newPermissions = append(newPermissions, newPermission)
+		}
+	}
+
+	if len(newPermissions) == 0 {
+		return false, nil
+	}
+
+	_, err = self.ec2.AuthorizeSecurityGroupIngress(securityGroupId, newPermissions)
+	if err != nil {
+		glog.Warning("error authorizing security group ingress", err)
+		return false, err
+	}
+
+	return true, nil
+}
+
+// CreateTCPLoadBalancer implements TCPLoadBalancer.CreateTCPLoadBalancer
+func (self *AWSCloud) CreateTCPLoadBalancer(name, region string, externalIP net.IP, ports []int, hosts []string, affinity api.AffinityType) (string, error) {
+	glog.V(2).Infof("CreateTCPLoadBalancer(%v, %v, %v, %v, %v)", name, region, externalIP, ports, hosts)
+
+	if affinity != api.AffinityTypeNone {
+		// ELB supports sticky sessions, but only when configured for HTTP/HTTPS
+		return "", fmt.Errorf("unsupported load balancer affinity: %v", affinity)
+	}
+
+	if len(externalIP) > 0 {
+		return "", fmt.Errorf("External IP cannot be specified for AWS ELB")
+	}
+
+	instances, err := self.getInstancesByDnsNames(hosts)
+	if err != nil {
+		return "", err
+	}
+
+	vpc, err := self.findVPC()
+	if err != nil {
+		glog.Error("error finding VPC", err)
+		return "", err
+	}
+	if vpc == nil {
+		return "", fmt.Errorf("Unable to find VPC")
+	}
+
+	// Construct list of configured subnets
+	subnetIds := []string{}
+	{
+		subnets, err := self.ec2.DescribeSubnets(nil, vpc.VpcId)
+		if err != nil {
+			return "", err
+		}
+
+		//	zones := []string{}
+		for _, subnet := range subnets {
+			subnetIds = append(subnetIds, subnet.SubnetId)
+			if !strings.HasPrefix(subnet.AvailabilityZone, region) {
+				glog.Error("found AZ that did not match region", subnet.AvailabilityZone, " vs ", region)
+				return "", fmt.Errorf("invalid AZ for region")
+			}
+			//		zones = append(zones, subnet.AvailabilityZone)
+		}
+	}
+
+	// Build the load balancer itself
+	var loadBalancerName, dnsName string
+	{
+		loadBalancer, err := self.describeLoadBalancer(region, name)
+		if err != nil {
+			return "", err
+		}
+
+		if loadBalancer == nil {
+			createRequest := &elb.CreateLoadBalancer{}
+			// TODO: Does k8s keep a UUID for the load balancer, which we could use instead of
```

Comment:
We don't, but perhaps we should. What do you need the UUID for that the cluster name + namespace + service name doesn't handle?